src/core/exec-credential.c

```diff
@@ -473,20 +473,30 @@ static int maybe_decrypt_and_write_credential(
         assert(data || size == 0);
 
         if (args->encrypted) {
+                CredentialFlags flags = 0; /* only allow user creds in user scope */
+
                 switch (args->params->runtime_scope) {
 
                 case RUNTIME_SCOPE_SYSTEM:
-                        /* In system mode talk directly to the TPM */
-                        r = decrypt_credential_and_warn(
-                                        id,
-                                        now(CLOCK_REALTIME),
-                                        /* tpm2_device= */ NULL,
-                                        /* tpm2_signature_path= */ NULL,
-                                        getuid(),
-                                        &IOVEC_MAKE(data, size),
-                                        CREDENTIAL_ANY_SCOPE,
-                                        &plaintext);
-                        break;
+                        /* In system mode talk directly to the TPM – unless we live in a device sandbox
+                         * which might block TPM device access. */
+
+                        flags |= CREDENTIAL_ANY_SCOPE;
+
+                        if (!args->context->private_devices) {
```

Comment:
Hmm, I figure this alone is not a complete fix, because DevicePolicy= exists. And the check of exec_context_has_encrypted_credentials() in unit_patch_context has lost relevance since the introduction of ImportCreds= as we can't tell for sure whether we need to decrypt it. Hence maybe go via IPC when any constraint is placed on device access, and drop mangling of device allowlist?